### PR TITLE
Laboratory work #1, Vyahireva Sasha - 19FPL1

### PR DESCRIPTION
**Pull request by micefic**

### CODE REVIEW

**Comment by ValeriyaKuznetsova** [comment]
Unfortunately, the deadline is not that far. Please, push changes so that we can review them.

File: lab_1/main.py [hunk 1]
```diff
@@ -12,6 +12,15 @@ def tokenize(text: str) -> list:
     e.g. text = 'The weather is sunny, the man is happy.'
     --> ['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy']
     """
+
+
+    def tokenize(text: str):
+        punctuation = ['(', ')', '?', ':', ';', ',', '.', '!', '/', '"', "'"]
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: what if there is '<' or '>'? 
either add all possible punctuation marks or use another way to check punctuation (more universal)

File: lab_1/main.py [hunk 2]
```diff
@@ -12,6 +12,15 @@ def tokenize(text: str) -> list:
     e.g. text = 'The weather is sunny, the man is happy.'
     --> ['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy']
     """
+
+
+    def tokenize(text: str):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: why use "def" in another "def"? there has been already a function declaration (everywhere)

Comment (ValeriyaKuznetsova) [reply lab_1/main.py]: just remove it

File: lab_1/main.py [hunk 3]
```diff
@@ -25,6 +34,17 @@ def remove_stop_words(tokens: list, stop_words: list) -> list:
     stop_words = ['the', 'is']
     --> ['weather', 'sunny', 'man', 'happy']
     """
+
+    def remove_stop_words():
+        tokens = str(input('Введите предложение ')).split()
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: no inputs(), tokens and stop_words are the arguments of this function, they already exist

File: lab_1/main.py [hunk 4]
```diff
@@ -25,6 +34,17 @@ def remove_stop_words(tokens: list, stop_words: list) -> list:
...
+
+        stop_words = str(input('Введите стоп-слова ')).split()
+
+        final_list = [word for word in tokens if word not in stop_words]
+
+        print(final_list)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: do not forget to remove print() when you push something

File: lab_1/main.py [hunk 5]
```diff
@@ -36,6 +56,16 @@ def calculate_frequencies(tokens: list) -> dict:
     e.g. tokens = ['weather', 'sunny', 'man', 'happy']
     --> {'weather': 1, 'sunny': 1, 'man': 1, 'happy': 1}
     """
+
+    def analysis():
+        tokens = str(input('Введите предложение ')).split()
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: same, no inputs

File: lab_1/main.py [hunk 6]
```diff
@@ -12,6 +12,15 @@ def tokenize(text: str) -> list:
...
+
+
+    def tokenize(text: str):
+        punctuation = ['(', ')', '?', ':', ';', ',', '.', '!', '/', '"', "'"]
+        for i in punctuation:
+            text = text.replace(i, "")
+
+    text = text.lower()
+    return text.split()
     pass
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: if there is something written in a function, remove 'pass'

File: lab_1/main.py [hunk 7]
```diff
@@ -12,17 +12,16 @@ def tokenize(text: str) -> list:
...
-            text = text.replace(i, "")
+    for i in text:
+        if i in punctuation:
+            text = text.replace(i, '')
 
     text = text.lower()
-    return text.split()
-    pass
-
+    text = text.split()
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: can you join lines 22 and 23?

File: lab_1/main.py [hunk 8]
```diff
@@ -34,18 +33,14 @@ def remove_stop_words(tokens: list, stop_words: list) -> list:
...
     """
+    pass
 
-    def remove_stop_words():
-        tokens = str(input('Введите предложение ')).split()
-
-        stop_words = str(input('Введите стоп-слова ')).split()
-
-        final_list = [word for word in tokens if word not in stop_words]
+    with open("stop_words.txt") as file:  # конструкция with...as позволяет не закрывать файл вручную
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: you shouldn't read any files in main.py
This is for start.py

stop_words is an argument of the remove_stop_words function
consider that it exists already

File: lab_1/main.py [hunk 9]
```diff
@@ -56,18 +51,17 @@ def calculate_frequencies(tokens: list) -> dict:
...
     pass
 
+    freq_dict = {}
+
+    for i in text1:
+        if i in freq_dict:
+            freq_dict[i] += 1
+        else:
+            freq_dict[i] = 1
+    return (freq_dict)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: remove brackets

File: lab_1/main.py [hunk 10]
```diff
@@ -12,8 +12,16 @@ def tokenize(text: str) -> list:
...
     """
-    pass
+    tokens = ''
+    punctuation = set('!"#$%&\'()*+,-./:;<=>?@[\]^_`{|}~.')
+
+    for i in tokens:
+        if i in punctuation:
+            tokens = tokens.replace(i, '')
+    tokens = tokens.lower()
+    tokens = tokens.split()
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: can you join lines 21 and 22?

File: lab_1/main.py [hunk 11]
```diff
@@ -69,7 +97,23 @@ def get_concordance(tokens: list, word: str, left_context_size: int, right_conte
     right_context_size = 3
     --> [['man', 'is', 'happy', 'the', 'dog', 'is'], ['dog', 'is', 'happy', 'but', 'the', 'cat']]
     """
-    pass
+    concordance = []
+    if type(tokens) != list or type(word) != str:
+        return []
+    elif type(left_context_size) != int or type(right_context_size) != int:
+        return []
+    elif left_context_size < 1 and right_context_size < 1:
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: can you join lines 101, 103, 105?

File: lab_1/main.py [hunk 12]
```diff
@@ -105,7 +159,11 @@ def write_to_file(path_to_file: str, content: list):
     """
     Writes the result in a file
     """
-    pass
+    with open(path_to_file, 'w', encoding='utf-8') as file:
+        for context in content:
+            context = ' '.join(context)
+            context += '\n'
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: can you first get a full context, not lines and only then write to file?

File: lab_1/main.py [hunk 13]
```diff
@@ -69,7 +92,23 @@ def get_concordance(tokens: list, word: str, left_context_size: int, right_conte
...
+        return []
+
+    indexes = [i for i, x in enumerate(tokens) if x == word]
+    if left_context_size > 0 and right_context_size > 0:
+        [concordance.append(tokens[i - left_context_size: i + right_context_size + 1]) for i in indexes]
+    elif left_context_size > 0:
+        [concordance.append(tokens[i - left_context_size: i + 1]) for i in indexes]
+    elif right_context_size > 0:
+        [concordance.append(tokens[i: i + right_context_size + 1]) for i in indexes]
+# can i not change these lines? it works anyway.
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: not these, I meant if-checks of types (now they are on lines 96,98,100)

File: lab_1/main.py [hunk 14]
```diff
@@ -87,16 +126,26 @@ def get_adjacent_words(tokens: list, word: str, left_n: int, right_n: int) -> li
     right_n = 3
     --> [['man', 'is'], ['dog, 'cat']]
     """
-    pass
+    func_tokens = get_adjacent_words(tokens, word, left_n, right_n)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: are you sure you need recursion here? maybe you meant get_concordance?

**Review by ValeriyaKuznetsova — COMMENTED** [review]
correct the issues and work with tests

**Comment by ValeriyaKuznetsova** [comment]
do not forget about start.py

File: lab_1/start.py [hunk 15]
```diff
@@ -1,18 +1,35 @@
...
+    print("concordance for 'team':", concordance[:3])
+
+    adjacent_words = main.get_adjacent_words(tokens, 'team', 1, 2)
+    print("adjacent words for word 'team':", adjacent_words[:3])
 
-    RESULT = None
+    RESULT = concordance
     # DO NOT REMOVE NEXT LINE - KEEP IT INTENTIONALLY LAST
-    assert RESULT == [(), ()], 'Concordance not working'
+    assert RESULT == [['time', 'a'], ['time', 'years'], ['olympic', 'great']], 'Concordance not working'
```
Comment (ValeriyaKuznetsova) [inline lab_1/start.py]: result is concordance, but after == you have put adjacent words
that's why start.py fails

Comment (ValeriyaKuznetsova) [reply lab_1/start.py]: write 
assert RESULT == [expected result]

File: lab_1/main.py [hunk 16]
```diff
@@ -12,7 +14,11 @@ def tokenize(text: str) -> list:
     e.g. text = 'The weather is sunny, the man is happy.'
     --> ['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy']
     """
-    pass
+    tokens = []
+    if text and isinstance(text, str):
+        tokens = text.lower()
+        tokens = re.sub('[^a-zA-Z \n]', '', tokens).split()
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: if text.lower, why use A-Z?

**Comment by ValeriyaKuznetsova** [comment]
target score should be 10

**Comment by ValeriyaKuznetsova** [comment]
Do not put "Review required" if you haven't corrected all the issues

**Comment by ValeriyaKuznetsova** [comment]
correct the issue about start.py

File: lab_1/main.py (hunk 16 above)
Comment (ValeriyaKuznetsova) [reply lab_1/main.py]: \w doesn't help :)
it's a-zA-Z0-9

**Comment by micefic** [comment]
> correct the issue about start.py



> correct the issue about start.py

what kind of issue?

**Comment by ValeriyaKuznetsova** [comment]
I've written them earlier
 
write
assert RESULT == [expected result]

put there concordance or sorted one, whatever

**Comment by ValeriyaKuznetsova** [comment]
Open another PR with the second lab

**Pull request closed without merging**